iaso/api/workflows.py

```diff
@@ -0,0 +1,199 @@
+
+def make_deep_copy_with_relations(orig_version):
+    orig_changes = WorkflowChange.objects.filter(workflow_version=orig_version)
+    orig_follow_ups = WorkflowFollowup.objects.filter(workflow_version=orig_version)
+
+    new_version = deepcopy(orig_version)
+    new_version.id = None
+    new_version.name = "Copy of " + orig_version.name
+    new_version.status = WorkflowVersionsStatus.DRAFT
+    new_version.save()
+
+    for oc in orig_changes:
+        new_change = deepcopy(oc)
+        new_change.workflow_version = new_version
+        new_change.id = None
+        new_change.save()
+
+    for of in orig_follow_ups:  # Doesn't copy the forms !
+        new_followup = deepcopy(of)
+        new_followup.workflow_version = new_version
+        new_followup.id = None
+        new_followup.save()
+
+        new_followup.forms.clear()
+
+        orig_forms = of.forms.all()
+        new_followup.forms.add(*orig_forms)
+
+    return new_version
+
+
+entity_type_id_param = openapi.Parameter(
+    "entity_type_id", openapi.IN_QUERY, description="Entity Type ID", type=openapi.TYPE_STRING, required=True
+)
+
+
+class WorkflowVersionViewSet(ModelViewSet):
+    """Workflow API
+    GET /api/workflowversion/
+    GET /api/workflowversion/{version_id}/
+    If version_id is provided returns the detail of this workflow version.
+    Else returns a paginated list of all the workflow versions.
+    """
+
+    permission_classes = [permissions.IsAuthenticated, HasPermission("menupermissions.iaso_workflows")]  # type: ignore
+    filter_backends = [filters.OrderingFilter, DjangoFilterBackend]
+    ordering_fields = ["name", "created_at", "updated_at"]
+    serializer_class = WorkflowVersionDetailSerializer
+    results_key = "workflow_versions"
+    remove_results_key_if_paginated = False
+    model = WorkflowVersion
+
+    lookup_url_kwarg = "version_id"
+
+    filterset_fields = {"workflow__entity_type": ["exact"]}
+
+    http_method_names = ["get", "post"]
+
+    @swagger_auto_schema(request_body=no_body)
+    @action(detail=True, methods=["post"])
+    def copy(self, request, **kwargs):
+        """POST /api/workflowversion/{version_id}/copy
+        Creates a new workflow version by copying the exiting version given by {version_id}
+        """
+
+        version_id = request.query_params.get("version_id", kwargs.get("version_id", None))
+        wv_orig = WorkflowVersion.objects.get(pk=version_id)
+        new_vw = make_deep_copy_with_relations(wv_orig)
+        serialized_data = WorkflowVersionSerializer(new_vw).data
+        return Response(serialized_data)
+
+    @swagger_auto_schema(manual_parameters=[entity_type_id_param], request_body=no_body)
+    def create(self, request, *args, **kwargs):
+        """POST /api/workflowversion/?entity_type_id=XXX
+        Create a new empty and DRAFT workflow version for the workflow connected to Entity Type 'entity_type_id'
+        """
+        entity_type_id = request.query_params.get("entity_type_id", None)
```

Comment:
I suggest we move the parameter `entity_type_id` to the body using a serializer instead. Actually if we just don't specify the create I think it will works by itself.

We should check if the user has access to the entity but we can do that in the validator of the Serializer instead.

Reply:
@olethanh can you check my modifications regarding this ?